### PR TITLE
feat(RBAC): RHICOMPL-3851 restrict host access to inventory groups

### PR DESCRIPTION
**Pull request by skateman**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

**Comment by skateman** [comment]
Sporiadic failures addressed by https://github.com/RedHatInsights/compliance-backend/pull/1755

File: app/models/host.rb [hunk 1]
```diff
@@ -125,6 +134,20 @@ def self.available_os_versions
...
+  def self.arel_inventory_groups(groups, key)
+    jsons = groups.map { |group| [{ key => group }].to_json.dump }
+
+    return Arel.sql('1 = 0') if jsons.empty?
+
+    Arel::Nodes::InfixOperation.new(
+      '@>', arel_table[:groups],
+      Arel::Nodes::NamedFunction.new(
+        'ANY',
+        [Arel::Nodes::SqlLiteral.new("'{#{jsons.join(',')}}'::jsonb[]")]
```
[the remark this answers is not in the document]
Comment (skateman) [reply app/models/host.rb:146]: I am building these JSONs 😉 from the RBAC response that is parsed by @romanblanco

File: app/models/host.rb [hunk 2]
```diff
@@ -117,6 +119,13 @@ class Host < ApplicationRecord
     joins("LEFT OUTER JOIN (#{sq.to_sql}) sq ON sq.id = hosts.id")
   }
 
+  scope :with_groups, lambda { |groups, key = :id|
+    # Skip the [] representing ungrouped hosts from the array when generating the query
+    grouped = arel_inventory_groups(groups.flatten, key)
+    # The OR is inside of Arel in order to prevent pollution of already applied scopes
+    where(groups.include?([]) ? grouped.or(UNGROUPED_HOSTS) : grouped)
```
[the remark this answers is not in the document]
Comment (skateman) [reply app/models/host.rb:126]: Empty array means access to ungrouped hosts cc @romanblanco

[the remark this answers is not in the document]
Comment (vkrizan) [reply app/models/host.rb:126]: I see. @romanblanco  Maybe we can a comment on all possibilities for `User#inventory_groups` method, including the ANY (`*`).

File: app/models/user.rb [hunk 3]
```diff
@@ -7,17 +7,21 @@ class User < ApplicationRecord
...
 
     rbac_permissions.any? do |access|
       Rbac.verify(access.permission, access_request)
     end
   end
 
   def inventory_groups
+    # We should not fetch inventory groups if the RBAC feature is globally disabled
+    # or the request is authenticated via an entitlement certificate.
+    return Rbac::ANY if rbac_disabled? || cert_authenticated?
```
Comment (vkrizan) [inline app/models/user.rb:23]: Shouldn't we restrict all inventory groups (return empty list) in case of cert auth?

**Review by vkrizan — APPROVED** [review]

**Pull request merged**